### PR TITLE
Implements Auth provider configuration management APIs

### PR DESCRIPTION
**Pull request by bojeil-google**
Implements Auth provider configuration management APIs for configuring SAML and OIDC providers.

```
listProviderConfigs(
  options: admin.auth.AuthProviderConfigFilter
): Promise<admin.auth.ListProviderConfigResults>;
getProviderConfig(providerId: string): Promise<admin.auth.AuthProviderConfig>
deleteProviderConfig(providerId: string): Promise<void>;
updateProviderConfig(
  providerId: string, updatedConfig: admin.auth.UpdateAuthProviderRequest
): Promise<admin.auth.AuthProviderConfig>;
createProviderConfig(
  config: admin.auth.AuthProviderConfig
): Promise<admin.auth.AuthProviderConfig>;
```

### CODE REVIEW

File: test/integration/auth.spec.ts [hunk 1]
```diff
@@ -425,6 +428,269 @@ describe('admin.auth', () => {
...
+    });
+
+    it('createProviderConfig() successfully creates a SAML config', () => {
+      return admin.auth().createProviderConfig(authProviderConfig1)
+        .then((config) => {
+          assertDeepEqualUnordered(authProviderConfig1, config);
+        });
+    });
+
+    it('getProviderConfig() successfully returns the expected SAML config', () => {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/integration/auth.spec.ts]: I have moved the creation of the common auth provider to the `before` block here and below. However, I think for integration tests reliant on an external state/remote server, it is ok to run tests in a specific order. Note we are dependent on order in the rest of the tests.
I can restructure all the tests to run in one giant single test as they do depend on each other and I do not want to make expensive and repetitive network requests before each test to set up the actual test scenario. However, that means losing granularity of testing one API at a time.

File: test/integration/auth.spec.ts [hunk 2]
```diff
@@ -425,6 +428,269 @@ describe('admin.auth', () => {
...
+    });
+
+    it('createProviderConfig() successfully creates an OIDC config', () => {
+      return admin.auth().createProviderConfig(authProviderConfig1)
+        .then((config) => {
+          assertDeepEqualUnordered(authProviderConfig1, config);
+        });
+    });
+
+    it('getProviderConfig() successfully returns the expected OIDC config', () => {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/integration/auth.spec.ts]: Same as above.

File: src/auth/auth-api-request.ts [hunk 3]
```diff
@@ -496,12 +504,139 @@ const FIREBASE_AUTH_GET_OOB_CODE = new ApiSettings('/accounts:sendOobCode', 'POS
...
+      throw new FirebaseAuthError(AuthClientErrorCode.INVALID_PAGE_TOKEN);
+    }
+    // Validate max results.
+    if (!validator.isNumber(request.pageSize) ||
+        request.pageSize <= 0 ||
+        request.pageSize > MAX_LIST_PROVIDER_CONFIGURATION_PAGE_SIZE) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        `Required "maxResults" must be a positive integer that does not exceed ` +
+        `the allowed ${MAX_LIST_PROVIDER_CONFIGURATION_PAGE_SIZE}.`,
```
Comment (hiranya911) [inline src/auth/auth-api-request.ts]: Remove `the allowed`

File: src/auth/auth-config.ts [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+  public readonly providerId: string;
+  public readonly idpEntityId: string;
+  public readonly ssoURL: string;
+  public readonly x509Certificates: string[];
+  public readonly rpEntityId: string;
+  public readonly callbackURL?: string;
+  public readonly enableRequestSigning?: boolean;
+
+  /**
+   * Static method convert a client side request to a SAMLConfigServerRequest.
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: Just `Converts a client side request to a SAMLConfigServerRequest, which is the format accepted by the backend server.`.

File: src/auth/auth-config.ts [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+   *
+   * @param {SAMLAuthProviderRequest} options The options object to convert to a server request.
+   * @param {boolean=} ignoreMissingFields Whether to ignore missing fields.
+   * @return {?SAMLConfigServerRequest} The resulting server request or null if not valid.
+   */
+  public static buildServerRequest(
+      options: SAMLAuthProviderRequest,
+      ignoreMissingFields: boolean = false): SAMLConfigServerRequest | null {
+    const makeRequest = typeof options !== 'undefined' &&
+        validator.isNonNullObject(options) &&
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: I think you can drop the check for undefined in the previous line, since this check captures it.

File: src/auth/auth-config.ts [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+      return null;
+    }
+    return matchProviderRes[1];
+  }
+
+  /**
+   * @param {any} providerId The provider ID to check.
+   * @return {boolean} Whether the provider ID corresponds to a SAML provider.
+   */
+  public static isProviderId(providerId: any): boolean {
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: Declare return type with a type guard on `providerId`.

`public static isProviderId(providerId: any): providerId is string {`

This way, once this method is called, `providerId` can be handled as a string at the callsite.

File: src/auth/auth-config.ts [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+    return validator.isNonEmptyString(providerId) && providerId.indexOf('saml.') === 0;
+  }
+
+  /**
+   * Validates the SAMLConfig options object. Throws an error on failure.
+   *
+   * @param {SAMLAuthProviderRequest} options The options object to validate.
+   * @param {boolean=} ignoreMissingFields Whether to ignore missing fields.
+   */
+  public static validate(options: SAMLAuthProviderRequest, ignoreMissingFields: boolean = false) {
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: Can this be private?

File: src/auth/auth-config.ts [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+      return null;
+    }
+    return matchProviderRes[1];
+  }
+
+  /**
+   * @param {any} providerId The provider ID to check.
+   * @return {boolean} Whether the provider ID corresponds to an OIDC provider.
+   */
+  public static isProviderId(providerId: any): boolean {
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: Add type guard

File: src/auth/auth-config.ts [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+    return validator.isNonEmptyString(providerId) && providerId.indexOf('oidc.') === 0;
+  }
+
+  /**
+   * Validates the OIDCConfig options object. Throws an error on failure.
+   *
+   * @param {OIDCAuthProviderRequest} options The options object to validate.
+   * @param {boolean=} ignoreMissingFields Whether to ignore missing fields.
+   */
+  public static validate(options: OIDCAuthProviderRequest, ignoreMissingFields: boolean = false) {
```
Comment (hiranya911) [inline src/auth/auth-config.ts]: This PR is a lot of code, and it's not easy to figure out which should be public and which not. Lets make sure we are only exposing things that we really need.

File: test/resources/mocks.ts [hunk 10]
```diff
@@ -167,6 +167,20 @@ export let keyPairs = [
   /* tslint:enable:max-line-length */
 ];
 
+// Randomly generate an X.509 certs using https://www.samltool.com/self_signed_certs.php
```
Comment (hiranya911) [inline test/resources/mocks.ts]: Randomly generated X.505 certs using...

File: test/unit/auth/auth.spec.ts [hunk 11]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+      stubs = [];
+    });
+
+    it('should be rejected given no provider ID', () => {
+      return (auth as any).getProviderConfig()
+        .should.eventually.be.rejected.and.have.property('code', 'auth/invalid-provider-id');
+    });
+
+    it('should be rejected given an invalid provider ID', () => {
+      const invalidProviderId = '';
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Perhaps also test for non-string args

File: test/unit/auth/auth.spec.ts [hunk 12]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+        pageToken: 'NEXT_PAGE_TOKEN',
+      };
+      const emptyListConfigsResponse: any = {
+        oauthIdpConfigs: [],
+      };
+      const emptyExpectedResult: any = {
+        providerConfigs: [],
+      };
+
+      it('should resolve on listOAuthIdpConfigs request success with configs in response', () => {
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: I'd rephrase this as `should resolve on success with configs in response`. Similar changes elsewhere.

Basically, `listOAuthIdpConfigs` is an internal implementation detail that we have to mock out, and there's no need to mention it everywhere.

File: test/unit/auth/auth.spec.ts [hunk 13]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+        // Stub listOAuthIdpConfigs to return expected response.
+        const listConfigsStub = sinon
+          .stub(FirebaseAuthRequestHandler.prototype, 'listOAuthIdpConfigs')
+          .returns(Promise.resolve(listConfigsResponse));
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: `.resolves(listConfigsResponse)`. Here and elsewhere.

File: test/unit/auth/auth.spec.ts [hunk 14]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+            expect(listConfigsStub)
+              .to.have.been.calledOnce.and.calledWith(maxResults, pageToken);
+          });
+      });
+
+      it('should throw an error when listOAuthIdpConfigs returns an error', () => {
+        // Stub listOAuthIdpConfigs to throw a backend error.
+        const listConfigsStub = sinon
+          .stub(FirebaseAuthRequestHandler.prototype, 'listOAuthIdpConfigs')
+          .returns(Promise.reject(expectedError));
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: `.rejects(expectedError)`. Here and elsewhere.

File: test/unit/auth/auth.spec.ts [hunk 15]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+    it('should be rejected given an invalid AuthProviderConfigFilter type', () => {
+      const invalidOptions = {
+        type: 'unsupported',
+      };
+      return (auth as Auth).listProviderConfigs(invalidOptions as any)
+        .then(() => {
+          throw new Error('Unexpected success');
+        })
+        .catch((error) => {
+          expect(error).to.have.property('code', 'auth/argument-error');
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Can't we do `.should.eventually.be.rejected.and.have.property` here?

File: test/unit/auth/auth.spec.ts [hunk 16]
```diff
@@ -1915,6 +1964,839 @@ describe('Auth', () => {
...
+        pageToken: 'NEXT_PAGE_TOKEN',
+      };
+      const emptyListConfigsResponse: any = {
+        inboundSamlConfigs: [],
+      };
+      const emptyExpectedResult: any = {
+        providerConfigs: [],
+      };
+
+      it('should resolve on listInboundSamlConfigs request success with configs in response', () => {
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: should resolve on success with configs in response

File: test/unit/auth/auth-api-request.spec.ts [hunk 17]
```diff
@@ -2404,6 +2408,1017 @@ describe('FirebaseAuthRequestHandler', () => {
     });
   });
 
+  describe('getOAuthIdpConfig()', () => {
+    const providerId = 'oidc.provider';
+    const path = `/v2beta1/projects/project_id/oauthIdpConfigs/${providerId}`;
+    const method = 'GET';
```
Comment (hiranya911) [inline test/unit/auth/auth-api-request.spec.ts]: Nit: Calling this `getMethod` or `httpGet` would make the rest of the code more readable. Or even using the string literal as `callParams(path, 'GET', {})` is not a bad idea.

**Comment by hiranya911** [comment]
As for managing large changes, this is what I would recommend.

1. Create a branch for the new feature and enable branch protection for it (you can ask me or another Firebase org admin to do this for you).

```
git checkout -b auth-id-providers master
// enable protection for auth-id-providers
``` 

2. Then create a series of pull requests against the new branch. Branch protection will make sure that PRs have to be reviewed and CI checks must pass before they can be merged.

```
git checkout -b impl-part1 auth-idp-providers
// create PR, review and merge

git checkout -b impl-part2 auth-idp-providers
// create PR, review and merge
```

3. When the entire feature is implemented, we can merge the new branch to master. This can be done in a PR from the new branch to the master.
